client-index/src/auto_synchronizer.rs

```diff
@@ -80,40 +80,74 @@ impl AutoSynchronizer {
         assert!(self.core.is_some());
     }
 
+    // to close channel, all senders of the channel should be dropped
+    // currently it has two senders
+    // 1. stream
+    // 2. sending queue
+
+    fn close_connection(&self) {
+        let mut data = self
+            .send_queue
+            .lock()
+            .expect("autosync close connection send-queue lock");
+        match &data.queue {
+            Some(a) => {
+                drop(a);
+                data.queue = None;
+            }
+            None => {}
+        }
+    }
+
+    fn process_text(&self, a: &String) -> std::result::Result<(), ()> {
+        let j: serde_json::Value = serde_json::from_str(&a).map_err(|_e| {})?;
+        if j["error"].is_null() {
+            self.close_connection();
+            if let Some(core) = self.core.as_ref() {
+                core.send(OwnedMessage::Text(a.clone())).map_err(|_e| {})?;
+            }
+        }
+        Ok(())
+    }
+
     /// activate tokio websocket
     pub fn run_network(&mut self) -> Result<()> {
         loop {
+            let mut connected = false;
             let channel = futures::sync::mpsc::channel(0);
             // tx, rx
             let (channel_tx, channel_rx) = channel;
             {
                 let mut data = self.send_queue.lock().unwrap();
                 data.queue = Some(channel_tx.clone());
             }
-            log::info!("Connecting to {}", self.websocket_url);
             let mut runtime = tokio::runtime::current_thread::Builder::new()
                 .build()
                 .expect("get tokio builder");
             // get synchronous sink
             let mut channel_sink = channel_tx.clone().wait();
+            drop(channel_tx);
```

Comment:
It looks ok, but is there any way to avoid doing a manual drop? Maybe move the code in a block to create a new scope and the values will get automatically dropped when the variable goes out of scope.

Reply:
good idea, i'll experiment

Reply:
these drop are used in scope, so using drop is necessary to remove all sender references.
but i removed unnecessary drop in other part.(found by clippy)